### ISSUE TITLE
Some of transaction_test cases crush under stress conditions consistently

### ISSUE THREAD

**Issue by yuslepukhin** [facebook/rocksdb#3701]
> Note: Please use Issues only for bug reports. For questions, discussions, feature requests, etc. post to dev group: https://www.facebook.com/groups/rocksdb.dev

### Actual behavior

The following is consistently exhibited by the following test cases under stress (concurrent run of many test cases). Dumps indicate that either the database or the root database pointers point to a memory that was filled by jemalloc as deallocated memory.

StackableDBAsBaseDB/TransactionTest.SeqAdvanceTest/1
StackableDBAsBaseDB/TransactionTest.SeqAdvanceTest/0
StackableDBAsBaseDB/TransactionTest.DuplicateKeys/1
 

WARNING: transaction_test_je::StackableDBAsBaseDB/TransactionTest.SeqAdvanceTest/1 State: Completed
WARNING: Note: Google Test filter = StackableDBAsBaseDB/TransactionTest.SeqAdvanceTest/1
WARNING: [==========] Running 1 test from 1 test case.
WARNING: [----------] Global test environment set-up.
WARNING: [----------] 1 test from StackableDBAsBaseDB/TransactionTest
WARNING: [ RUN      ] StackableDBAsBaseDB/TransactionTest.SeqAdvanceTest/1
WARNING: **unknown file: error: SEH exception with code 0xc0000005 thrown in the test body.**
WARNING: unknown file: error: SEH exception with code 0xc0000005 thrown in the test fixture's destructor.
WARNING: [  FAILED  ] StackableDBAsBaseDB/TransactionTest.SeqAdvanceTest/1, where GetParam() = (true, true, 1) (19312 ms)
WARNING: [----------] 1 test from StackableDBAsBaseDB/TransactionTest (19312 ms total)
WARNING:
WARNING: [----------] Global test environment tear-down
WARNING: [==========] 1 test from 1 test case ran. (19313 ms total)
WARNING: [  PASSED  ] 0 tests.
WARNING: [  FAILED  ] 1 test, listed below:
WARNING: [  FAILED  ] StackableDBAsBaseDB/TransactionTest.SeqAdvanceTest/1, where GetParam() = (true, true, 1)
WARNING:
WARNING:  1 FAILED TEST

### Steps to reproduce the behavior
Run transaction_test test cases in parallel against a RAM drive. I am surprised this is not caught by tools.

**Comment by yuslepukhin** [comment]
It looks like on the tenth iteration the RootDb of the Stackable becomes uninitialized and points to a deal located memory. This is StackableDBAsBaseDB/TransactionTest.SeqAdvanceTest/0

![image](https://user-images.githubusercontent.com/11303988/38646585-6bfcdb1a-3d9d-11e8-9361-5bcbefa856b6.png)

**Comment by maysamyabandeh** [comment]
Thanks @yuslepukhin for reporting. I unfortunately could not reproduce this. Can you try with this patch and see what the returned status is when it crashes?

$ git diff
diff --git a/utilities/transactions/transaction_test.cc b/utilities/transactions/transaction_test.cc
index 8b34717..4584501 100644
--- a/utilities/transactions/transaction_test.cc
+++ b/utilities/transactions/transaction_test.cc
@@ -4946,7 +4946,9 @@ TEST_P(TransactionTest, SeqAdvanceTest) {
     return n & filter;
   };
   const size_t max_n = static_cast<size_t>(1) << NUM_BRANCHES;
-  for (size_t n = 0; n < max_n; n++, ReOpen()) {
+  Status s;
+  for (size_t n = 0; n < max_n; n++, s = ReOpen()) {
+    assert(s.ok());
     DBImpl* db_impl = reinterpret_cast<DBImpl*>(db->GetRootDB());
     size_t branch = 0;
     auto seq = db_impl->GetLatestSequenceNumber();

**Comment by maysamyabandeh** [comment]
Thanks @yuslepukhin. Looking forward to the PR then.

**Comment by yuslepukhin** [comment]
@maysamyabandeh

This makes use `--gtest_filter=MySQLStyleTransactionTest/MySQLStyleTransactionTest.TransactionStressTest/6 --gtest_catch_exceptions=0` as an example.

With my adjustments about 40% of the test cases crashes. Here is why.

Without RTTI static_cast_with_check does not check for anything when casting from StackableDB to DBImpl.  And the RTTI is apparently off in debug builds by default. 

![image](https://user-images.githubusercontent.com/11303988/38698852-4ba919b0-3e4b-11e8-9b0a-e4b1cc36149e.png)

PessimisticTransactionDB __ctor then proceeds to treat StackableDB as DBImpl and point to unallocated memory beyond the object. But StackableDB does not implement DBImpl it contains it as a RootDB().

![image](https://user-images.githubusercontent.com/11303988/38699321-8bde8f28-3e4c-11e8-9c5f-c629f8bbcae4.png)

The funny and ironic  part that `WritePreparedTxnDB::Initialize()` is also somewhat confused as to what needs to be used. The first half of the function makes use of dbimpl obtained from GetRootDB(). Then the second half of the function switches to db_impl_ which is a protected member of one of the base classes (!) and this is where the trouble starts where it attemps to access DBImpl::mutex_

![image](https://user-images.githubusercontent.com/11303988/38699679-8955255e-3e4d-11e8-9ac0-f575fbcdac06.png)

The issues from WrapStackableDB() and WrapDB() has to do with the unclear failure semantics and leaking database objects. There is a number of uninitialized ptrs that makes debugging harder and the test carefully avoids those issues but here is what I am seeing:

- Typically, when database Open() calls fail the user does not expect to clean up anything. Here and below https://github.com/facebook/rocksdb/blob/master/utilities/transactions/pessimistic_transaction_db.cc#L260 The failure return causes the passed db ptr to leak because the burden of deleting it is on the user. 
- But the user is not always clear that it needs to delete it because in the few lines down the ownership of the passed db is taken by WritePreparedTxnDB which passes it to StackableDB https://github.com/facebook/rocksdb/blob/master/utilities/transactions/pessimistic_transaction_db.cc#L262 
- At the point the output `dbptr` is also initialized. The problem is that it all happens before potential Intiialize() failure.
- So if the initialize fails the user ends up with two pointers it must cleanup, one of them owns another so the possibility of double delete is all too clear.
- In one case the caller must delete the initial database, in another it must delete the wrapper but not the `db` it passes as an argument. How can anyone get this right? That is, of course, if the user is aware that the burden of the clean is on them because the usual convention that the nothing is returned. Thus `TransactionDB::Open()` simply leaks either the first db which was opened by `DBImpl::Open()` or two dbs if the `Initialize()` fails

There are too many issues with this code, I am not sure I have time for the task to fix all the issues as I want to change the majority of the code  (I feel young again! :) ) We will need to postpone the upgrade.

**Comment by maysamyabandeh** [comment]
@yuslepukhin Thanks for debugging this; and sorry for the delayed response; github is dropping notifications here and there.
> And the RTTI is apparently off in debug builds by default.

How did that happen? In debug mode RTTI must be enable: https://github.com/facebook/rocksdb/blob/3be9b364539d382ce282cad80b5c05c58fd2d8a7/Makefile#L124
> when casting from StackableDB to DBImpl

PessimisticTransactionDB has two constructors: one for db_impl and one for [stackable_db](https://github.com/facebook/rocksdb/blob/3be9b364539d382ce282cad80b5c05c58fd2d8a7/utilities/transactions/pessimistic_transaction_db.cc#L66). The one that you [linked](https://github.com/facebook/rocksdb/blob/3be9b364539d382ce282cad80b5c05c58fd2d8a7/utilities/transactions/pessimistic_transaction_db.cc#L35) is for db_impl. What was the bug that would use the wrong constructor?

Thanks for pointing out the initialization issues. Will surely work with you to fix them. But I still do not know what is the bug that makes your tests failing, which I cannot reproduce!

**Comment by yuslepukhin** [comment]
@maysamyabandeh Thanks for responding.

You are correct about the wrong constructor. I have introduced this bug while fixing the leak that described above and put a wrong type for the smart_ptr on line 295 in my patch. That type should be `StackableDB`

With regards to RTTI. There are two issues in the below 3 lines of code :)
1) That macro is not defined in Cmake file. After I defined it it started asserting.
2) This is a wrong macro to use because it creates an option to disable a valuable diagnostic facility in debug builds. IMHO, the correct line is `#ifdef _DEBUG`
![image](https://user-images.githubusercontent.com/11303988/39377139-03626e12-4a09-11e8-932c-019ea3fa4c9e.png)

**Comment by maysamyabandeh** [comment]
1. I hope we migrate to a unified build system soon to avoid such issues. Can you kindly send a PR for the fix for CMake builds?
2. I believe that was the intention: to keep dynamic_cast in debug mode by default. If that is not the case, it would a bug:
>  130 * Replace dynamic_cast<> (except unit test) so people can choose to build with RTTI off. With make, release mode is by default built with -fno-rtti and debug mode is built wi    thout it. Users can override it by setting USE_RTTI=0 or 1.

**Comment by yuslepukhin** [comment]
@maysamyabandeh 1) My fix was temp, just adding `ROCKSDB_USE_RTTI` macro to the compiler flags for MS compiler as I do not have facility to test it on other platforms. Additional compiler flags on other platforms may be needed for other platforms to enable rtti, we do not disable it.
2) Generally, I agree with what you are saying bc it is a well known pattern. However, the quoted comment does not make sense to me. You can't just disable rtti in the unit tests, it has to be on everywhere especially bc it is used in the library itself. So the correct macro should be _DEBUG there. If someone wants to build/run debug w/o rtti then that function can not be implemented. At least it can not be used as intended and called as such.

**Comment by siying** [comment]
@yuslepukhin it was my mistake to have overlooked Cmake when introducing the ROCKSDB_USE_RTTI macro. The default behavior of Cmake should be enabling RTTI. Now it is wrong.

**Comment by gfosco** [comment]
Please let me know if #3801 is going in the right direction to address this.

**Comment by gfosco** [comment]
Closing, can re-open if needed.
